# Assessing the power of deep generative models for learning joint embeddings of single-cell multi-omics data
This is the repository that accompanies the paper "Assessing the power of deep generative models for learning joint embeddings of single-cell multi-omics data."

## Abstract
Recent extensions of single-cell studies to multiple data modalities raise new questions regarding
experimental design. Specifically, as these methods obtain more information from each cell,
one might expect multi-omics approaches to require fewer cells than single-omics methods. For
example, the challenge of sparsity in single-omics data might be resolved by compensating
for missing information across modalities. With deep learning approaches, particularly deep
generative models (DGMs), emerging as a promising method for integrating these complex
datasets, investigations into sample size requirements become highly relevant. Here, we
empirically examine the quality of DGM-based integrations for varying sample sizes. Therefore,
we first review the literature and give a short overview of deep learning methods for multi-
omics integration. Next, we use six popular tools to examine their robustness to different cell
numbers, where we consider two of the most common multi-omics types at the moment. We
use simultaneous gene expression measurements at the RNA level and protein abundance
measurements for cell surface proteins (CITE-seq). In addition, we use data where chromatin
accessibility and RNA expression are measured in thousands of cells (10x Multiome). We
examine the methods regarding their ability to learn joint embeddings, where we subdivide the
evaluation into biological and technical metrics. Finally, we make recommendations for the design
of multi-omics experiments and discuss potential future developments.

## Data  

## Training  

## Plotting  

## References  
